server/app/tasks/DatabaseSeedTask.java

```diff
@@ -0,0 +1,162 @@
+              /* questionText= */ LocalizedStrings.of(
+                  Lang.forCode("en-US").toLocale(),
+                  "Please enter your date of birth in the format mm/dd/yyyy"),
+              /* questionHelpText= */ LocalizedStrings.empty()));
+
+  private final Provider<QuestionService> questionServiceProvider;
+  private final Provider<VersionRepository> versionRepository;
+  private final Database database;
+
+  @Inject
+  public DatabaseSeedTask(
+      Provider<QuestionService> questionService, Provider<VersionRepository> versionRepository) {
+    this.questionServiceProvider = checkNotNull(questionService);
+    this.versionRepository = checkNotNull(versionRepository);
+    this.database = DB.getDefault();
+
+    this.run();
+  }
+
+  private void run() {
+    if (isMissingCanonicalQuestions()) {
+      // Ensure a draft version exists to avoid transaction collisions.
+      versionRepository.get().getDraftVersion();
+      inSerializableTransaction(this::seedCanonicalQuestions, 1);
+    }
+  }
+
+  /**
+   * Ensures that questions with names matching those in {@code CANONICAL_QUESTIONS} are present in
+   * the database, inserting the definitions in {@code CANONICAL_QUESTIONS} if any aren't found.
+   */
+  private void seedCanonicalQuestions() {
+    var questionService = questionServiceProvider.get();
+    ImmutableSet<String> existingQuestionNames = questionService.getQuestionNames();
+
+    for (QuestionDefinition questionDefinition : CANONICAL_QUESTIONS) {
+      if (existingQuestionNames.contains(questionDefinition.getName())) {
+        LOGGER.info(
+            "Canonical question \"%s\" exists at server start", questionDefinition.getName());
+        continue;
+      }
+
+      ErrorAnd<QuestionDefinition, CiviFormError> result =
+          questionService.create(questionDefinition);
+
+      if (result.isError()) {
+        String errorMessages =
+            result.getErrors().stream()
+                .map(CiviFormError::message)
+                .collect(Collectors.joining(", "));
+
+        LOGGER.error(
+            String.format(
+                "Unable to create canonical question \"%s\" due to %s",
+                questionDefinition.getName(), errorMessages));
+      } else {
+        LOGGER.info("Created canonical question \"%s\"", questionDefinition.getName());
+      }
+    }
+  }
+
+  private void inSerializableTransaction(Runnable fn, int tryCount) {
+    Transaction transaction =
+        database.beginTransaction(TxScope.requiresNew().setIsolation(TxIsolation.SERIALIZABLE));
+
+    try {
+      fn.run();
+      transaction.commit();
+    } catch (NonUniqueResultException | SerializableConflictException | RollbackException e) {
+      LOGGER.warn("Database seed transaction failed: %s", e);
+
+      if (tryCount > MAX_RETRIES) {
+        throw new RuntimeException(e);
+      }
+
+      transaction.end();
+      inSerializableTransaction(fn, ++tryCount);
```

Comment:
Related to the above. Rather than rescheduling, could we Thread.sleep for a random backoff interval? Or is this occurring on the main thread?

Retry-in-a-tight-loop patterns generally make me wary.

Reply:
I believe it is happening on the main thread, which, per your question above, would block the server from becoming healthy. I'm not 100% on the behavior there so I'm going to experiment with it some and let you know.